pkg/pool-weighted/contracts/LBPool.sol

```diff
@@ -0,0 +1,312 @@
+    function _getNormalizedWeight(uint256 tokenIndex) internal view virtual override returns (uint256) {
+        if (tokenIndex < _NUM_TOKENS) {
+            return _getNormalizedWeights()[tokenIndex];
+        }
+
+        revert IVaultErrors.InvalidToken();
+    }
+
+    function _getNormalizedWeights() internal view override returns (uint256[] memory) {
+        uint256[] memory normalizedWeights = new uint256[](_NUM_TOKENS);
+        normalizedWeights[0] = _getNormalizedWeight0();
+        normalizedWeights[1] = FixedPoint.ONE - normalizedWeights[0];
+
+        return normalizedWeights;
+    }
+
+    function _getNormalizedWeight0() internal view virtual returns (uint256) {
+        PoolState memory poolState = _poolState;
+        uint256 pctProgress = GradualValueChange.calculateValueChangeProgress(poolState.startTime, poolState.endTime);
+
+        return GradualValueChange.interpolateValue(poolState.startWeight0, poolState.endWeight0, pctProgress);
+    }
+
+    function _getPoolSwapEnabled() private view returns (bool) {
+        return _poolState.swapEnabled;
+    }
+
+    function _setSwapEnabled(bool swapEnabled) private {
+        _poolState.swapEnabled = swapEnabled;
+
+        emit SwapEnabledSet(swapEnabled);
+    }
+
+    /**
+     * @dev When calling updateWeightsGradually again during an update, reset the start weights to the current weights,
+     * if necessary.
+     */
+    function _startGradualWeightChange(
+        uint32 startTime,
+        uint32 endTime,
+        uint256[] memory startWeights,
+        uint256[] memory endWeights
+    ) internal virtual {
+        PoolState memory poolState = _poolState;
+
+        poolState.startTime = startTime;
+        poolState.endTime = endTime;
+
+        // These have been validated, but SafeCast anyway out of an abundance of caution.
+        poolState.startWeight0 = startWeights[0].toUint64();
+        poolState.endWeight0 = endWeights[0].toUint64();
+
+        _poolState = poolState;
+
+        emit GradualWeightUpdateScheduled(startTime, endTime, startWeights, endWeights);
+    }
+
+    /**
+     * @dev Ensure the given set of weights sums to exactly FixedPoint.ONE, and neither of the weights is below
+     * the minimum.
+     */
+    function _ensureValidWeights(uint256[] memory normalizedWeights) internal pure {
+        InputHelpers.ensureInputLengthMatch(_NUM_TOKENS, normalizedWeights.length);
+
+        // Ensure each normalized weight is above the minimum
+        uint256 normalizedSum = 0;
+        for (uint8 i = 0; i < _NUM_TOKENS; ++i) {
+            uint256 normalizedWeight = normalizedWeights[i];
+
+            if (normalizedWeight < _MIN_WEIGHT) {
+                revert MinWeight();
+            }
+            normalizedSum += normalizedWeight;
+        }
+
+        // Ensure that the normalized weights sum to ONE
+        if (normalizedSum != FixedPoint.ONE) {
+            revert NormalizedWeightInvariant();
+        }
+    }
```

Comment:
Why get rid of the 2-token optimized function? Also if it's in a library like it was originally, it could be used from any weighted pool, reducing duplicated code among different Weighted(style)Pool variants. 

For example, vanilla WeightedPool, LBPool, (a hypothetical 2-token optimized v3) WeightedPool2Tokens could all use the same library.

For extra context, I had put it the (single-use) function in a library contract since I didn't want to overstep and change the vanilla WeightedPool contract to use it w/o discussing with y'all

Reply:
I just considered that pre-optimizing, which we tend to avoid (i.e., adding new libraries or functions that *might* be used in the future). The main weighted pool has immutable weights (as will probably most variants), so this would only be called once on deployment.

Ideally this could go in the base WeightedPool contract and just be called from here, but it can't because that pool is also storing immutables in the middle. We could do it anyway, and just do two loops in `WeightedPool`:

Instead of:
```
uint256 normalizedSum = 0;
for (uint8 i = 0; i < _totalTokens; ++i) {
    uint256 normalizedWeight = params.normalizedWeights[i];

    if (normalizedWeight < _MIN_WEIGHT) {
        revert MinWeight();
    }
    normalizedSum = normalizedSum + normalizedWeight;

    // prettier-ignore
    if (i == 0) { _normalizedWeight0 = normalizedWeight; }
    else if (i == 1) { _normalizedWeight1 = normalizedWeight; }
    else if (i == 2) { _normalizedWeight2 = normalizedWeight; }
    else if (i == 3) { _normalizedWeight3 = normalizedWeight; }
    else if (i == 4) { _normalizedWeight4 = normalizedWeight; }
    else if (i == 5) { _normalizedWeight5 = normalizedWeight; }
    else if (i == 6) { _normalizedWeight6 = normalizedWeight; }
    else if (i == 7) { _normalizedWeight7 = normalizedWeight; }
}

// Ensure that the normalized weights sum to ONE
if (normalizedSum != FixedPoint.ONE) {
    revert NormalizedWeightInvariant();
}
```

we'd have `_ensureValidWeights` defined in `WeightedPool`, and the code above would be:

```
_ensureValidWeights(params.normalizedWeights);

for (uint8 i = 0; i < _totalTokens; ++i) {
    uint256 normalizedWeight = params.normalizedWeights[i];

    // prettier-ignore
    if (i == 0) { _normalizedWeight0 = normalizedWeight; }
    else if (i == 1) { _normalizedWeight1 = normalizedWeight; }
    else if (i == 2) { _normalizedWeight2 = normalizedWeight; }
    else if (i == 3) { _normalizedWeight3 = normalizedWeight; }
    else if (i == 4) { _normalizedWeight4 = normalizedWeight; }
    else if (i == 5) { _normalizedWeight5 = normalizedWeight; }
    else if (i == 6) { _normalizedWeight6 = normalizedWeight; }
    else if (i == 7) { _normalizedWeight7 = normalizedWeight; }
}
```

Reply:
Ok fair enough to bring it back into the pool contract if it's not useful for WPs then.

Considering that an LBP (or other similar dynamic pools) could have multiple weight shifts over its lifetime, are you against an in-pool-contract version of the weight validator fn that uses the 2-token version? IMO it was pretty simple, clean code. Copying here w/o comments for easy reference:
```
function validateTwoWeights(uint256 normalizedWeight0, uint256 normalizedWeight1) internal pure {
    if (normalizedWeight0 < _MIN_WEIGHT || normalizedWeight1 < _MIN_WEIGHT) {
        revert MinWeight();
    }
    if (normalizedWeight0 + normalizedWeight1 != FixedPoint.ONE) {
        revert NormalizedWeightInvariant();
    }
}
```

Reply:
Given that this is only called in one place (because the base contract WeightedPool handles it on construction), I can just move this code inline. That's even simpler, and optimized.